pulser-core/pulser/backend/state.py

```diff
@@ -0,0 +1,151 @@
+
+Eigenstate = Union[States, Literal["0", "1"]]
+
+ArgScalarType = TypeVar("ArgScalarType")
+ReturnScalarType = TypeVar("ReturnScalarType")
+StateType = TypeVar("StateType", bound="State")
+
+
+class State(ABC, Generic[ArgScalarType, ReturnScalarType]):
+    """Base class enforcing an API for quantum states.
+
+    Each backend will implement its own type of state and the
+    methods below.
+    """
+
+    _eigenstates: Sequence[Eigenstate]
+
+    @property
+    @abstractmethod
+    def n_qudits(self) -> int:
+        """The number of qudits in the state."""
+        pass
+
+    @property
+    def eigenstates(self) -> tuple[Eigenstate, ...]:
+        """The eigenstates that form a qudit's eigenbasis.
+
+        The order of the states should match the order in a
+        numerical (ie state vector or density matrix)
+        representation.
+        """
+        return tuple(self._eigenstates)
+
+    @property
+    def qudit_dim(self) -> int:
+        """The dimensions (ie number of eigenstates) of a qudit."""
+        return len(self.eigenstates)
+
+    @abstractmethod
+    def overlap(self: StateType, other: StateType, /) -> ReturnScalarType:
+        """Compute the overlap between this state and another of the same type.
+
+        Generally computes Tr[AB] for mixed states A and B, which
+        corresponds to |<a|b>|^2 for pure states A=|a><a| and B=|b><b|.
+
+        Args:
+            other: The other state.
+
+        Returns:
+            The overlap between the two states.
+        """
+        pass
+
+    @abstractmethod
+    def sample(
+        self,
+        *,
+        num_shots: int,
+        one_state: Eigenstate | None = None,
+        p_false_pos: float = 0.0,
+        p_false_neg: float = 0.0,
+    ) -> Counter[str]:
+        """Sample bitstrings from the state, taking into account error rates.
+
+        Args:
+            num_shots: How many bitstrings to sample.
+            one_state: The eigenstate that measures to 1. Can be left undefined
+                if the state's eigenstates form a known eigenbasis with a
+                defined "one state".
+            p_false_pos: The rate at which a 0 is read as a 1.
+            p_false_neg: The rate at which a 1 is read as a 0.
+
+        Returns:
+            The measured bitstrings, by count.
+        """
+        pass
+
+    @classmethod
+    @abstractmethod
+    def from_state_amplitudes(
+        cls: Type[StateType],
+        *,
+        eigenstates: Sequence[Eigenstate],
+        amplitudes: dict[str, ArgScalarType],
+    ) -> StateType:
+        """Construct the state from its basis states' amplitudes.
+
+        Args:
+            eigenstates: The basis states (e.g., ('r', 'g')).
+            amplitudes: A mapping between basis state combinations and
+                complex amplitudes.
+
+        Returns:
+            The state constructed from the amplitudes.
+        """
+        pass
+
+    def get_correlation_matrix(
```

Comment:
Based on my monkey patching idea, we could remove this. Having it is a bit of a slipperly slope, since why not also add a `get_magnetization` and whatever other items you could think of.

Reply:
Sure, I'm fine with that